Update pom structure

mezpahlan opened the pull request:
The current POM structure is quite out of date to what you get when creating an up to date Hello World plugin from the tutorial web pages. This PR updates the structure and versions of the parent POM.

Along with being aligned to the new structure we also resolve Enforcer and Findbug errors that come as a result of the newer Jenkins plugin dependencies.

The overall aim was to add support for a friendly syntax whilst using this plugin in pipeline mode. Getting the POM up to date seems a good first step to that.

mezpahlan commented:
I see now. Yes, you're quite right. Apologies. (First time contributing to Jenkins, forgive me).

1. What would you suggest with regards to this PR? Undo the charges that break binary compatibility, undo changes that duplicate #45, leave the Findbugs fixes? Something else?
2. With regards to that breaks binary compatibility comment: Findbugs complains that the field the parameter sets is unused. So removing the field to make Findbugs happy surely means removing the parameter? 

Anyways, thanks for looking at this.

oleg-nenashev commented:
@mezpahlan no problem at all, it happens. Generally we need a plugin maintainer to review and integrate something. I would not mind if your PR gets integrated instead of mine. In #45 I have also replaced bundled libs by Apache HttpComponents Client 4.x API Plugin, but it can be done in a follow-up if needed.

I am not sure who is the current plugin maintainer, @Brantone @dkochetkov and @ohoeltke are listed in https://github.com/jenkins-infra/repository-permissions-updater/blob/master/permissions/plugin-hockeyapp.yml

> With regards to that breaks binary compatibility comment: Findbugs complains that the field the parameter sets is unused. So removing the field to make Findbugs happy surely means removing the parameter?

The field is public, so it *may* be used somehow. My recommendation would be to just use `@SupressFBWarnings` with a "TODO" comment for now

mezpahlan commented:
Thanks I've reverted that binary breaking change and added your suggestions. Hopefully one of the maintainers can also chime in and move this forward. I'd love to help out with further improvements to this plugin in the future too. Happy to help where I can.

oleg-nenashev commented on pom.xml:
```diff
@@ -1,134 +1,118 @@
+    <properties>
+        <jenkins.version>1.609.2</jenkins.version>
+        <java.level>7</java.level>
+        <jenkins-test-harness.version>2.38</jenkins-test-harness.version>
```
I would suggest to remove it and let O=PluginPOM use it's default one

oleg-nenashev commented on pom.xml:
```diff
@@ -1,134 +1,118 @@
+    <packaging>hpi</packaging>
+
+    <properties>
+        <jenkins.version>1.609.2</jenkins.version>
```
If you require Java 7, it's better to update to Jenkins 1.625.3. Just because 1.609.x can run on Java 6 and hence the plugin may not work there

oleg-nenashev commented on src/main/java/hockeyapp/HockeyappApplication.java:
```diff
@@ -28,11 +29,14 @@
+    @SuppressFBWarnings(value = {"URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD"}, justification = "Breaks binary compatibility if removed.")
     @XStreamAsAttribute
-    public long schemaVersion;
+    public long schemaVersion; // TODO: Fix Findbugs gracefully.
```
Should it be marked as deprecated then? Just in case thee is really no usages anywhere (including Jelly and Co)

oleg-nenashev commented on src/main/java/hockeyapp/HockeyappApplication.java:
```diff
@@ -146,25 +150,31 @@ public String getDisplayName() {
-                    return FormValidation.errorWithMarkup(
-                            "You must enter an <a href=\"https://rink.hockeyapp.net/manage/auth_tokens\">API Token</a>.");
+            if (value.isEmpty()) {
+                final Jenkins instance = Jenkins.getInstance();
```
in 1.625.3 you can use `getActiveInstance()` IIRC. Jenkins instance cannot be really null here

oleg-nenashev commented on src/main/java/hockeyapp/HockeyappRecorder.java:
```diff
@@ -82,7 +80,8 @@
 
-    private static final Charset UTF8_CHARSET = Charset.forName("UTF-8");
+    private static final String UTF8 = "UTF-8";
+    private static final Charset UTF8_CHARSET = Charset.forName(UTF8);
```
In Java 7 you can use `StandardCharsets`.

oleg-nenashev commented on pom.xml:
```diff
@@ -1,134 +1,118 @@
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>4.5.2</version>
```
Once the core is updated to 1.625.3, you could use Apache HttpComponents Client 4.x API plugin instead

oleg-nenashev commented on pom.xml:
```diff
@@ -1,134 +1,118 @@
+        </dependency>
+        <dependency>
+            <groupId>commons-codec</groupId>
+            <artifactId>commons-codec</artifactId>
```
Why don't you just use the version from the core?

oleg-nenashev commented on pom.xml:
```diff
@@ -1,134 +1,101 @@
+    </developers>
+
+
+    <url>http://wiki.jenkins-ci.org/display/JENKINS/HockeyApp+Plugin</url>
```
https://wiki.jenkins.io/....
